include/sleipnir/control/OCPSolver.hpp

```diff
@@ -0,0 +1,205 @@
+#include "sleipnir/optimization/OptimizationProblem.hpp"
+
+namespace sleipnir {
+
+/**
+ * Function representing an explicit or implicit ODE, or a discrete state
+ * transition function. Explicit: dx/dt = f(t, x, u) Implicit: f(t, [x dx/dt]',
+ * u) = 0 State transition: xₖ₊₁ = f(t, xₖ, u)
+ */
+using DynamicsFunction = std::function<VariableMatrix(
+    double, const VariableMatrix&, const VariableMatrix&)>;
+
+/**
+ * Constrain a fixed step OCP. This function is called numSteps + 1 times, once
+ * for each step. The arguments are X_i (numStates)x1, U_i (numInputs)x1, and
+ * then true if this is the last step and has no associated input, otherwise
+ * false.
+ */
+using FixedStepConstraintFunction =
+    std::function<void(const VariableMatrix&, const VariableMatrix&)>;
+
+/**
+ * Performs 4th order Runge-Kutta integration of dx/dt = f(t, x, u) for dt.
+ *
+ * @param f  The function to integrate. It must take two arguments x and u.
+ * @param x  The initial value of x.
+ * @param u  The value u held constant over the integration period.
+ * @param t0 The initial time.
+ * @param dt The time over which to integrate.
+ */
+template <typename F, typename State, typename Input, typename Time>
+State RK4(F&& f, State x, Input u, Time t0, Time dt) {
+  auto halfdt = dt * 0.5;
+  State k1 = f(t0, x, u);
+  State k2 = f(t0 + halfdt, x + halfdt * k1, u);
+  State k3 = f(t0 + halfdt, x + halfdt * k2, u);
+  State k4 = f(t0 + dt, x + dt * k3, u);
+
+  return x + (k1 + k2 * 2.0 + k3 * 2.0 + k4) * (dt / 6.0);
+}
+
+/**
+ * Enum describing an OCP transcription method.
+ */
+enum TranscriptionMethod {
+  kDirectTranscription,
+  kDirectCollocation,
+  kSingleShooting
+};
+
+/**
+ * Enum describing a type of system dynamics constraints.
+ */
+enum DynamicsType { kExplicitODE, kDiscrete };
+
+/**
+ * This class allows the user to pose and solve a constrained optimal control
+ * problem (OCP) in a variety of ways.
+ *
+ * The system is transcripted by one of three methods (direct collocation,
+ * direct transcription, or single-shooting) and additional constraints can be
+ * added.
+ *
+ * @todo expand this explanation more.
+ * In single-shooting, states depend explicitly as a function of all previous
+ * states and all previous inputs. In direct transcription, each state is a
+ * decision variable constrained to the integrated dynamics of the previous
+ * state. In direct collocation, the trajectory is modeled as a series of cubic
+ * polynomials where the centerpoint slope is constrained.
+ *
+ * Explicit ODEs are integrated using RK4.
+ *
+ * For explicit ODEs, the function must be in the form dx/dt = f(t, x, u).
+ * For discrete state transition functions, the function must be in the form
+ * xₖ₊₁ = f(t, xₖ, u).
+ *
+ * Direct collocation requires an explicit ODE. Direct transcription and
+ * single-shooting can use either an ODE or state transition function.
+ */
+class SLEIPNIR_DLLEXPORT FixedStepOCPSolver : public OptimizationProblem {
```

Comment:
Prefer composition over inheritance unless you want to expose the entire base class interface (seems unlikely here).

Reply:
I think I do want the entire base class interface exposed. `Minimize`, `Maximize`, `Solve`, and `SubjectTo` are all useful for this context and would just get passed through anyway. The only one I can't see much use for is `DecisionVariable` but I don't think there's harm in it.